### PR TITLE
Add Brave Leo P3A metrics

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves https://github.com/brave/brave-browser/issues/32050

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/ai_chat/browser/ai_chat_metrics.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_AI_CHAT_BROWSER_AI_CHAT_METRICS_H_
+#define BRAVE_COMPONENTS_AI_CHAT_BROWSER_AI_CHAT_METRICS_H_
+
+#include "base/memory/weak_ptr.h"
```
Comment (iefremov) [inline components/ai_chat/browser/ai_chat_metrics.h:9]: not used?

Comment (iefremov) [reply components/ai_chat/browser/ai_chat_metrics.h:9]: ah, I see `SupportsWeakPtr` but it is not needed for timers

File: components/ai_chat/browser/ai_chat_metrics.cc [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+AIChatMetrics::~AIChatMetrics() = default;
+
+void AIChatMetrics::RegisterPrefs(PrefRegistrySimple* registry) {
+  registry->RegisterListPref(prefs::kBraveChatP3AChatCountWeeklyStorage);
+  registry->RegisterListPref(prefs::kBraveChatP3APromptCountWeeklyStorage);
+}
+
+void AIChatMetrics::RecordEnabled() {
+  UMA_HISTOGRAM_BOOLEAN(kEnabledHistogramName, true);
```
Comment (iefremov) [inline components/ai_chat/browser/ai_chat_metrics.cc:46]: we do not want to emit false signal by default? So we are going to collect only `true`s and then look at the amount?

File: components/ai_chat/browser/ai_chat_metrics.cc [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+void AIChatMetrics::RecordNewChat() {
+  chat_count_storage_.AddDelta(1);
+}
+
+void AIChatMetrics::RecordNewPrompt() {
```
Comment (iefremov) [inline components/ai_chat/browser/ai_chat_metrics.cc:53]: this probably should be reviewed by ai chat codeowners to check logic

**Review by iefremov — APPROVED** [review]

File: components/ai_chat/browser/ai_chat_metrics.cc (hunk 2 above)
Comment (DJAndries) [reply components/ai_chat/browser/ai_chat_metrics.cc:46]: Yes, that's right. We will use the total samples from another question as the denominator for calculating the ratio. More details on this process can be found in the handbook that I wrote.

File: components/ai_chat/browser/ai_chat_metrics.cc (hunk 3 above)
Comment (DJAndries) [reply components/ai_chat/browser/ai_chat_metrics.cc:53]: cc @nullhook

File: components/ai_chat/browser/ai_chat_tab_helper.cc [hunk 4]
```diff
@@ -85,6 +88,9 @@ bool AIChatTabHelper::HasUserOptedIn() {
 
 void AIChatTabHelper::OnUserOptedIn() {
   MaybeGeneratePageText();
+  if (ai_chat_metrics_ != nullptr && HasUserOptedIn()) {
```
Comment (bridiver) [inline components/ai_chat/browser/ai_chat_tab_helper.cc:91]: nit: `!= nullptr` is not necessary

**Pull request merged**